File: PrivacySandboxSampleProject/SampleSandboxClient/app/src/main/java/com/example/client/MainActivity.java
/*
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.example.client;

import android.app.sdksandbox.SdkSandboxManager;
import android.app.sdksandbox.SdkSandboxManager.RemoteSdkCallback;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.SurfaceControlViewHost.SurfacePackage;
import android.view.SurfaceView;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import java.util.concurrent.Executor;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "SandboxClient";
    /**
     * Name of the SDK to be loaded.
     */
    private static final String SDK_NAME = "com.example.provider";
    /**
     * Button to load the SDK to the sandbox.
     */
    private Button mLoadSdkButton;
    /**
     * Button to request a SurfacePackage from sandbox which remotely render a webview.
     */
    private Button mRequestWebViewButton;
    /**
     * An instance of SdkSandboxManager which contains APIs to communicate with the sandbox.
     */
    private SdkSandboxManager mSdkSandboxManager;
    /**
     * The SurfaceView which will be used by the client app to show the SurfacePackage
     * going to be rendered by the sandbox.
     */
    private SurfaceView mClientView;

    private boolean mSdkLoaded = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mSdkSandboxManager = getApplicationContext().getSystemService(
            SdkSandboxManager.class);

        mClientView = findViewById(R.id.rendered_view);
        mClientView.setZOrderOnTop(true);

        mLoadSdkButton = findViewById(R.id.load_sdk_button);
        mRequestWebViewButton = findViewById(R.id.request_webview_button);

        registerLoadCodeProviderButton();
        registerRequestWebViewButton();
    }

    /**
     * Register the callback action after once mLoadSdkButton got clicked.
     */
    private void registerLoadCodeProviderButton() {
        mLoadSdkButton.setOnClickListener(v -> {
          log("Attempting to load sandbox SDK");
          final RemoteSdkCallbackImpl callback = new RemoteSdkCallbackImpl();
          mSdkSandboxManager.loadSdk(
              SDK_NAME, new Bundle(), Runnable::run, callback);
        });
    }

    /**
     * Register the callback action after once mRequestWebViewButton got clicked.
     */
    private void registerRequestWebViewButton() {
        mRequestWebViewButton.setOnClickListener(v -> {
          if (!mSdkLoaded) {
            makeToast("Please load the SDK first!");
            return;
          }
          log("Getting SurfacePackage.");
          new Handler(Looper.getMainLooper()).post(() -> {
            Bundle bundle = new Bundle();
            mSdkSandboxManager.requestSurfacePackage(
                SDK_NAME, getDisplay().getDisplayId(),
                mClientView.getWidth(), mClientView.getHeight(), bundle);
              });
        });
    }

    /**
     * A callback for tracking events regarding loading and interacting with SDK.
     */
    private class RemoteSdkCallbackImpl implements RemoteSdkCallback {
        private RemoteSdkCallbackImpl() {}

        /**
         * This notifies client application that the requested SDK is successfully loaded.
         *
         * @param params list of params returned from Sdk to the App.
         */
        @Override
        public void onLoadSdkSuccess(Bundle params) {
            log("onLoadSdkSuccess: " + params);
            makeToast("Loaded successfully!");
            mSdkLoaded = true;
        }

        /**
         * This notifies client application that the requested Sdk is failed to be loaded.
         *
         * @param errorCode int code for the error
         * @param errorMessage a String description of the error
         */
        @Override
        public void onLoadSdkFailure(int errorCode, String errorMessage) {
            log("onLoadSdkFailure(" + errorCode + "): " + errorMessage);
            makeToast("Load SDK Failed!" + errorMessage);
        }

        /**
         * This notifies client application that {@link SurfacePackage}
         * is ready to remote render view from the SDK.
         *
         * @param surfacePackage the requested surface package by
         *            {@link SdkSandboxManager#requestSurfacePackage(String, int, int, int, Bundle)}
         * @param surfacePackageId a unique id for the {@link SurfacePackage} {@code surfacePackage}
         * @param params list of params returned from Sdk to the App.
         */
        @Override
        public void onSurfacePackageReady(SurfacePackage surfacePackage,
            int surfacePackageId, Bundle params) {
            log("Surface package ready: " + params);
            makeToast("Surface Package Rendered!");
            new Handler(Looper.getMainLooper()).post(() -> {
              log("Setting surface package in the client view");
              mClientView.setChildSurfacePackage(surfacePackage);
              mClientView.setVisibility(View.VISIBLE);
            });
        }

        /**
         * This notifies client application that requesting {@link SurfacePackage} has failed.
         *
         * @param errorCode int code for the error
         * @param errorMessage a String description of the error
         */
        @Override
        public void onSurfacePackageError(int errorCode, String errorMessage) {
            log("onSurfacePackageError" + errorCode + "): " + errorMessage);
            makeToast("Surface Package Failed!" + errorMessage);
        }
    }

    private void makeToast(String message) {
        runOnUiThread(() -> Toast.makeText(MainActivity.this, message, Toast.LENGTH_SHORT).show());
    }

    private void log(String message) {
        Log.e(TAG, message);
    }
}

File: PrivacySandboxKotlin/client-app/src/main/java/com/example/client/MainActivity.kt
/*
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.example.client

import android.os.Bundle
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.CheckBox
import android.widget.Spinner
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.example.privacysandbox.client.R
import com.runtimeaware.sdk.BannerAd
import com.runtimeaware.sdk.ExistingSdk
import com.runtimeaware.sdk.FullscreenAd
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity() {
    /** Container for rendering content from the SDK. */
    private lateinit var bannerAd: BannerAd

    private val runtimeAwareSdk = ExistingSdk(this)

    /** A spinner for selecting the size of the file created in the sandbox. */
    private lateinit var fileSizeSpinner: Spinner

    /** Represents a file size that can be selected in the UI. */
    private data class FileSize(val sizeInMb: Int) {
        /** Called when FileSize is shown in the spinner. */
        override fun toString() = "$sizeInMb MB"
    }

    private val fileSizes = listOf(
        FileSize(3),
        FileSize(9),
        FileSize(18),
    )

    /** A spinner for selecting the type of ad being requested. */
    private lateinit var adTypeSpinner: Spinner
    private val adTypes = listOf("Banner", "WebView Banner")

    /** A spinner for selecting the mediation option. */
    private lateinit var mediationDropDownMenu: Spinner

    // Mediation Option values.
    // Please keep the order here the same as the order in which the options occur in the
    // mediation_dropdown_menu_array.
    //
    // As SDKs transition into the SDK Runtime, we may have some SDKs still in the app process
    // while the mediator and other SDKs have moved.
    // RUNTIME_MEDIATEE Mediated Ads is the scenario when the winning ad network is Runtime Enabled
    // as is the Mediator.
    // INAPP_MEDIATEE Mediated Ads is the scenario when the winning ad network is running in the
    // same process as the app and the Mediator is Runtime Enabled.
    enum class MediationOption {
        NONE,
        RUNTIME_MEDIATEE,
        INAPP_MEDIATEE,
        REFRESH_MEDIATED_ADS
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        bannerAd = findViewById(R.id.banner_ad)

        findViewById<Button>(R.id.initialize_sdk_button).setOnClickListener {
            onInitializeSkButtonPressed()
        }
        findViewById<Button>(R.id.create_file_button).setOnClickListener {
            onCreateFileButtonPressed()
        }
        findViewById<Button>(R.id.request_banner_button).setOnClickListener {
            onRequestBannerButtonPressed()
        }
        findViewById<Button>(R.id.fullscreen_button).setOnClickListener {
            showFullscreenView()
        }

        fileSizeSpinner = findViewById<Spinner>(R.id.create_file_size_spinner).apply {
            adapter = ArrayAdapter(this@MainActivity,
                android.R.layout.simple_spinner_dropdown_item,
                fileSizes,
            )
        }

        adTypeSpinner = findViewById<Spinner>(R.id.request_ad_spinner).apply {
            adapter = ArrayAdapter(
                this@MainActivity, android.R.layout.simple_spinner_dropdown_item, adTypes)
        }

        mediationDropDownMenu = findViewById(R.id.mediation_options_dropdown)

        ArrayAdapter.createFromResource(
            this@MainActivity,
            R.array.mediation_dropdown_menu_array,
            android.R.layout.simple_spinner_item
        ).also { adapter ->
            adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            mediationDropDownMenu.adapter = adapter
        }

    }

    private fun onInitializeSkButtonPressed() = lifecycleScope.launch {
        if (!runtimeAwareSdk.initialize()) {
            makeToast("Failed to initialize SDK")
        } else {
            makeToast("Initialized SDK!")
        }
    }

    private fun onRequestBannerButtonPressed() = lifecycleScope.launch {
        // Apps can allow or deny activity launches as they happen. In this example we are
        // surfacing a checkbox that controls the launches. In production apps could disable
        // launches whenever they feel SDKs shouldn't be launching activities (in the middle of
        // certain game scenes, video playback, etc).
        val loadWebView = adTypes[adTypeSpinner.selectedItemPosition].contains("WebView")
        // Mediated Banner Ad is shown when RUNTIME_MEDIATEE Mediation option is chosen.
        val mediationType =
            MediationOption.entries[mediationDropDownMenu.selectedItemId.toInt()].toString()
        bannerAd.loadAd(
            this@MainActivity,
            PACKAGE_NAME,
            shouldStartActivityPredicate(),
            loadWebView,
            mediationType
        )
    }

    private fun showFullscreenView() = lifecycleScope.launch {
        val mediationType =
            MediationOption.entries[mediationDropDownMenu.selectedItemId.toInt()].toString()
        val fullscreenAd = FullscreenAd.create(this@MainActivity, mediationType)
        fullscreenAd.show(this@MainActivity, shouldStartActivityPredicate())
    }

    private fun shouldStartActivityPredicate() : () -> Boolean {
        return { findViewById<CheckBox>(R.id.sdk_activity_launch_checkbox).isChecked }
    }

    private fun onCreateFileButtonPressed() {
        val fileSize = fileSizes[fileSizeSpinner.selectedItemPosition]

        lifecycleScope.launch {
            val success = runtimeAwareSdk.createFile(fileSize.sizeInMb)
            if (success == null) {
                makeToast("Please load the SDK first!")
                return@launch
            }
            makeToast(success)
        }
    }

    private fun makeToast(message: String) {
        runOnUiThread { Toast.makeText(this@MainActivity, message, Toast.LENGTH_SHORT).show() }
    }

    companion object {
        private const val TAG = "SandboxClient"

        /**
         * Package name of this app. This is something that the SDK might use the identify this
         * particular app client.
         *
         * (Note that in this particular sample it's used to build the banner view label).
         */
        private const val PACKAGE_NAME = "com.example.privacysandbox.client"
    }
}
